lib/network_ssl/network_ssl.c

```diff
@@ -0,0 +1,640 @@
+
+/* SSL context in which to create connections. */
+static SSL_CTX * ctx = NULL;
+
+/* Internal state. */
+struct network_ssl_ctx {
+	/* SSL management state. */
+	int s;
+	SSL * ssl;
+	int waiting_r;
+	int waiting_w;
+	void * immediate_cookie;
+
+	/* Pending read operation. */
+	int (*read_callback)(void *, ssize_t);
+	void * read_cookie;
+	uint8_t * read_buf;
+	size_t read_buflen;
+	size_t read_minlen;
+	size_t read_bufpos;
+	int read_needs_r;
+	int read_needs_w;
+
+	/* Pending write operation. */
+	int (*write_callback)(void *, ssize_t);
+	void * write_cookie;
+	const uint8_t * write_buf;
+	size_t write_buflen;
+	size_t write_minlen;
+	size_t write_bufpos;
+	int write_needs_r;
+	int write_needs_w;
+};
+
+/* Try to SSL_read/SSL_write if possible. */
+static int poke(struct network_ssl_ctx *);
+
+/* Free the SSL context upon exit. */
+static void
+freectx(void)
+{
+
+	assert(ctx != NULL);
+	SSL_CTX_free(ctx);
+	ctx = NULL;
+}
+
+/* Initialize SSL library. */
+static int
+init(void)
+{
+	const SSL_METHOD * meth;
+
+	/* If we have a context already, there's nothing to do. */
+	if (ctx != NULL)
+		return (0);
+
+	/* Launch SSL. */
+	if (!SSL_library_init()) {
+		warn0("Could not initialize SSL library");
+		goto err0;
+	}
+
+	/* We want to use TLS. */
+	if ((meth = TLS_client_method()) == NULL) {
+		warn0("TLS_client_method");
+		goto err0;
+	}
+
+	/* Create an SSL context. */
+	if ((ctx = SSL_CTX_new(meth)) == NULL) {
+		warn0("Could not create SSL context");
+		goto err0;
+	}
+
+	/* Insist on a minimum of TLS 1.2. */
+	if (!SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION)) {
+		warn0("Could not set minimum TLS version");
+		goto err1;
+	}
+
+	/* Partial writes are a thing. */
+	SSL_CTX_set_mode(ctx, SSL_MODE_ENABLE_PARTIAL_WRITE);
+
+	/* Load root certificates. */
+	if (!SSL_CTX_set_default_verify_paths(ctx)) {
+		warn0("Could not load default root certificates");
+		goto err1;
+	}
+
+	/* Release the context when we exit. */
+	atexit(freectx);
+
+	/* Success! */
+	return (0);
+
+err1:
+	SSL_CTX_free(ctx);
+	ctx = NULL;
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Socket ready for reading. */
+static int
+callback_read(void * cookie)
+{
+	struct network_ssl_ctx * ssl = cookie;
+
+	/* No longer waiting for socket to be readable. */
+	ssl->waiting_r = 0;
+
+	/*
+	 * Any operations which were waiting for the socket to become
+	 * readable are now free to run.
+	 */
+	ssl->read_needs_r = 0;
+	ssl->write_needs_r = 0;
+
+	/* Poke the SSL stack. */
+	return (poke(ssl));
+}
+
+/* Socket ready for writing. */
+static int
+callback_write(void * cookie)
+{
+	struct network_ssl_ctx * ssl = cookie;
+
+	/* No longer waiting for socket to be writable. */
+	ssl->waiting_w = 0;
+
+	/*
+	 * Any operations which were waiting for the socket to become
+	 * writable are now free to run.
+	 */
+	ssl->read_needs_w = 0;
+	ssl->write_needs_w = 0;
+
+	/* Poke the SSL stack. */
+	return (poke(ssl));
+}
+
+/* Immediate callback from register or cancel functions. */
```

Comment:
While you're fixing things, can you `s/or cancel //` here?  I was originally planning on having cancel functions call this but changed my mind.

Reply:
done